cloudify_agent/api/plugins/installer.py

```diff
@@ -350,81 +351,83 @@ def _rmtree(path):
-        _patch_pip_download()
-        # Import here, after patch
-        import pip
-        pip.download.unpack_url(link=pip.index.Link(package_url),
-                                location=plugin_dir,
-                                download_dir=None,
-                                only_download=False)
-    except Exception as e:
+        command = [get_pip_path(), 'download', '-d',
+                   archive_dir, '--no-deps', package_url]
+        runner.run(command=command)
+        archive = _get_archive(archive_dir, package_url)
+        plugin_dir_parent = extract_archive(archive)
+        plugin_dir = _get_plugin_path(plugin_dir_parent, package_url)
+
+    except NonRecoverableError as e:
         if plugin_dir and os.path.exists(plugin_dir):
             shutil.rmtree(plugin_dir)
-        raise exceptions.PluginInstallationError(
-            'Failed to download and unpack package from {0}: {1}.'
-            'You may consider uploading the plugin\'s Wagon archive '
-            'to the manager, For more information please refer to '
-            'the documentation.'.format(package_url, str(e)))
+        raise e
+
     finally:
-        _restore_pip_download()
+        if os.path.exists(archive_dir):
+            shutil.rmtree(archive_dir)
 
     return plugin_dir
 
 
+def _get_plugin_path(plugin_dir_parent, package_url):
+    """
+    plugin_dir_parent is a directory containing the plugin dir. Since we
+    need the plugin`s dir, we find it's name and concatenate it to the
+    plugin_dir_parent.
+    """
+    contents = list(walk(plugin_dir_parent))
+    if len(contents) < 1:
+        _remove_tempdir_and_raise_proper_exception(package_url,
+                                                   plugin_dir_parent)
+    plugin_dir_name = contents[0][1][0]
+    return os.path.join(plugin_dir_parent, plugin_dir_name)
+
+
+def _assert_list_len_greater_than_value(l, value, package_url, archive_dir):
+    if len(l) < value:
+        _remove_tempdir_and_raise_proper_exception(package_url, archive_dir)
+
+
+def _assert_list_len(l, expected_len, package_url, archive_dir):
+    if len(l) != expected_len:
+        _remove_tempdir_and_raise_proper_exception(package_url, archive_dir)
+
+
+def _remove_tempdir_and_raise_proper_exception(package_url, tempdir):
+    if tempdir and os.path.exists(tempdir):
+        shutil.rmtree(tempdir)
+    raise exceptions.PluginInstallationError(
+        'Failed to download package from {0}.'
+        'You may consider uploading the plugin\'s Wagon archive '
+        'to the manager, For more information please refer to '
+        'the documentation.'.format(package_url))
+
+
+def _get_archive(archive_dir, package_url):
+    """
+    archive_dir contains a zip file with the plugin directory. This function
+    finds the name of that zip file and returns the full path to it (the full
+    path is required in order to extract it)
+    """
+    contents = list(walk(archive_dir))
```

Comment:
We're doing it twice. Let's have a `_get_dir_contents` function that does the walk, validates that the list has 1 element, and returns it, and use it both here and in `_get_plugin_path`.